modules/server/src/middlewares/xviz-websocket-sender.js

```diff
@@ -0,0 +1,153 @@
+    // - formatter
+
+    this.options = options;
+
+    // This is the actual format we use to send data and can change
+    // based on the message.
+    this.format = options.format;
+
+    if (this.format === XVIZFormat.OBJECT) {
+      // We can not output OBJECT on a websocket
+      this.format = XVIZFormat.BINARY_GLB;
+    }
+
+    this.writer = null;
+    // If format is undefined we want to send the 'natural' format of
+    // the data (as long as it's not an OBJECT).
+    // Keep track of current 'writer' format
+    this.writerFormat = null;
+
+    this._syncFormatWithWriter(this.format);
+  }
+
+  log(...msg) {
+    const {logger} = this.options;
+    logger.log(...msg);
+  }
+
+  // Sets this.writer based on 'format'
+  _syncFormatWithWriter(format) {
+    // Cover the case where we have a format and no writer or when the
+    // format does not match.
+    if (format && (!this.writer || this.writerFormat !== format)) {
+      this.writer = new XVIZFormatWriter(this.sink, {format});
+      this.writerFormat = format;
+    }
+  }
+
+  // Data is in the desired format and can be written to sink directly
+  _sendDataDirect(format, resp) {
+    const sourceFormat = resp.format;
+
+    // need to check if object() has been called (ie it might be dirty) and repack
+    if (format === sourceFormat && !resp.hasMessage()) {
+      return true;
+    }
+
+    return false;
+  }
+
+  // If the format is unspecified we output the 'natural' format
+  // if it is valid. Make that determination here.
+  _getFormatOptions(msg) {
+    // default should be pass-thru of original data
+    if (!this.format) {
+      // If no format is specified, we send the 'natural' format
+      // but it must be a string or arraybuffer, not an OBJECT
+
+      // Test to determine if msg is either string or arraybuffer
+      if (
+        msg.format === XVIZFormat.OBJECT ||
+        (!msg.hasMessage() && typeof msg.buffer !== 'string' && !msg.buffer.byteLength)
+      ) {
+        return XVIZFormat.BINARY;
+      }
+
+      // return the format set to the current data format
+      return msg.format;
+    }
+
+    return this.format;
+  }
+
+  onError(msg) {
+    // TODO: This message is almost always just a plain object
+    // but the special handling for here feels awkard
+    const response = JSON.stringify(msg.buffer);
+    this.sink.writeSync('error', response);
+  }
+
+  onMetadata(msg) {
+    const format = this._getFormatOptions(msg);
+
+    if (this._sendDataDirect(format, msg)) {
+      this.sink.writeSync(`1-frame`, msg.buffer);
+    } else {
+      this._syncFormatWithWriter(format);
+      this.writer.writeMetadata(msg);
+    }
+  }
+
+  onStateUpdate(msg) {
+    const format = this._getFormatOptions(msg);
+
+    if (this._sendDataDirect(format, msg)) {
+      this.sink.writeSync('2-frame', msg.buffer);
+    } else {
+      this._syncFormatWithWriter(format);
+      this.writer.writeFrame(0, msg);
+    }
+  }
+
+  onTransformLogDone(msg) {
```

Comment:
Yeah it is weird that some of the message we send back and forth as GLB but some we always send as JSON.

Reply:
I should just do the same I do for `onStateUpdate(msg) ` here.  I'm cheating because this is a server generated msg but yeah, this is not robust.

Reply:
curiously, in order to support the uniform handling of message I need the above `writeMessage` change :)  The reason being is that this code is taking  short cut by assuming it knows the data format and writing directly to the socket.  The more uniform approach delegates the formatting issue to the *writer*. But the current IO interface has write functions only for Metadata and Frame.   So this too will be in #454 which is turning out to be a very much needed correction on the API